### PR TITLE
Feature/form drafts publish flags

### PR DESCRIPTION
**Pull request by parc-jason**
# Description
Allow for drafts of form versions/schemas.

Allow for form having a published version (or not). Returns published=true/false, versionUpdatedAt is when version created/published

Add API endpoint to publish/unpublish version.
`/forms/formId/versions/versionId/publish(optional params to unpublish are ?unpublish=true)`

Add API endpoints to list, create, read, update, delete draft.
`/forms/formId/drafts`

Add API endpoint to publish draft.
`/forms/formId/drafts/draftId/publish`

Add API endpoint to fetch form with only a published version (if it has one).
`/forms/formId/version`

Adjust the My Forms/RBAC to provide publish flags and dates (published = true/false, versionUpdatedAt).

**IMPORTANT** I have 'deprecated' some API calls that interfere with the new logic.  They currently do not appear to be in use (in master).  POST /forms/id/versions, and PUT /forms/id/versions/vid will return 410.  We don't want create or update a version. We will create a version either at form create, or via publishing a draft.  The only update to a version will be publish or unpublish and that has a specific endpoint.

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-wxpbtr-tools.pathfinder.gov.bc.ca/dashboard?id=common-hosted-form-service-pr-46

File: app/src/db/migrations/20201008142046_006-version-drafts.js [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+};
+
+exports.down = function(knex) {
+  // OK, this is really ugly.
+  // One cannot simply remove a column from a view.
+  // One must drop the view and then recreate; however, one must drop ALL the dependant objects too, then recreate them.
+  // Very lame.
+  // first, drop the view that uses the new column (and all objects that use that view)
+  // then we drop the column from the table
+  // then we re-create the view and all the views that depend on it.
```
Comment (loneil) [inline app/src/db/migrations/20201008142046_006-version-drafts.js:51]: 😬

Comment (parc-jason) [reply app/src/db/migrations/20201008142046_006-version-drafts.js:51]: Haha, yup.  can only add columns to the end of the list (without dropping first).
This going to be a pain if we constantly update views... :(

**Review by loneil — APPROVED** [review]
Yeah I think this is good stuff. Not sure if you want to merge though with the PVC code in the DC.

File: app/src/forms/form/service.js [hunk 2]
```diff
@@ -110,45 +123,41 @@ const service = {
...
+      .modify('filterPublished', params.published)
       .modify('orderVersionDescending');
   },
 
-  createVersion: async (formId, data, currentUser) => {
+  publishVersion: async (formId, formVersionId, params = {}, currentUser) => {
     let trx;
     try {
+      // allow an unpublish if they pass in unpublish parameter with an affirmative
+      const publish = params.unpublish ? falsey(params.unpublish) : true;
```
[the remark this answers is not in the document]
Comment (parc-jason) [reply app/src/forms/form/service.js:137]: not really, but i've been through this before and write utility functions, so figured it was time to bite the bullet.
no dependencies, so figured it was worth it.

[the remark this answers is not in the document]
Comment (loneil) [reply app/src/forms/form/service.js:137]: kk

**Pull request merged**